lib/wibox/widget/graph.lua

```diff
@@ -142,19 +168,43 @@ function graph.draw(_graph, _, cr, width, height)
             end
         end
 
+        local x0 = cr:get_matrix().x0
+
         -- Draw the background on no value
         if #values ~= 0 then
             -- Draw reverse
             for i = 0, #values - 1 do
                 local value = values[#values - i]
                 if value >= 0 then
+                    local x = i*step_width + ((i-1)*step_spacing) + 0.5
                     value = (value - min_value) / max_value
-                    cr:move_to(i + 0.5, height * (1 - value))
-                    cr:line_to(i + 0.5, height)
+                    cr:move_to(x, height * (1 - value))
+
+                    if step_shape then
+                        cr:translate(step_width + (i>1 and step_spacing or 0), height * (1 - value))
+                        step_shape(cr, step_width, height)
+                        cr:translate(nil, -(height * (1 - value)))
+                    elseif step_width > 1 then
+                        cr:rectangle(x, height * (1 - value), step_width, height)
+                    else
+                        cr:line_to(x, height)
+                    end
                 end
             end
             cr:set_source(color(_graph._private.color or beautiful.graph_fg or "#ff0000"))
-            cr:stroke()
+
+            if step_shape or step_width > 1 then
+                cr:fill()
+
+                -- Undo all the translates
+                if step_shape then
+                    local mat = cr:get_matrix()
+                    mat.x0 = x0
+                    cr:set_matrix(mat)
```

Comment:
Uhm. Could you somehow use `:save()` / `:restore()` instead? For one thing: Because your code wont work inside of a rotating widget. For another thing: You aren't undoing *all*  the translations, but only the translation in x direction.